src/backend/distributed/utils/background_jobs.c

```diff
@@ -385,292 +629,347 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
+						 * concurrent modification. We first check if a task is concurrently cancelled
+						 * or deleted. Accordingly, we decide to make the task status as cancelled or running.
+						 */
+
+
+						/*
+						 * reload task while holding a new ExclusiveLock on the table. A separate process
+						 * could have cancelled or removed the task by now, they would not see the pid and
+						 * status update, so it is our responsibility to stop the backend and _not_ write
+						 * the pid and running status.
+						 *
+						 * The lock will release on transaction commit.
+						 */
+						LockRelationOid(DistBackgroundTaskRelationId(), ExclusiveLock);
+
+						task = GetBackgroundTaskByTaskId(handleEntry->taskid);
+
+						ereport(LOG, (errmsg("found task with jobid/taskid: %ld/%ld",
+											 task->jobid, task->taskid)));
+
+						if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING)
+						{
+							/*
+							 * being in that step means that a concurrent cancel happened. we should make task status
+							 * as cancelled. We also want to reflect cancel message by consuming task executor
+							 * queue.
+							 */
+							taskCancelledConcurrently = true;
+						}
+						else
+						{
+							/*
+							 * now that we have verified the task has not been cancelled and still exist we
+							 * update it to reflect the new state. If task is already in running status,
+							 * the operation is idempotent. But for runnable tasks, we make their status
+							 * as running.
+							 */
+
+							pid_t pid = 0;
+							GetBackgroundWorkerPid(handleEntry->handle, &pid);
+							task->status = BACKGROUND_TASK_STATUS_RUNNING;
+							SET_NULLABLE_FIELD(task, pid, pid);
+
+							/* Update task status to indicate it is running */
+							UpdateBackgroundTask(task);
+							UpdateBackgroundJob(task->jobid);
+						}
+
+						monitorExecutionState = TryConsumeTaskWorker;
+						break;
+					}
+
+					case TryConsumeTaskWorker:
+					{
+						/*
+						 * we consume task executor response queue.
+						 * possible response codes can lead us different steps below.
+						 */
+						bool hadError = false;
+						shm_mq_result mq_res = ReadFromExecutorQueue(handleEntry->seg,
+																	 handleEntry->message,
+																	 &hadError);
+
+						if (taskCancelledConcurrently)
+						{
+							/* we update task status as cancelled */
+							ereport(LOG, (errmsg(
+											  "task jobid/taskid is cancelled: %ld/%ld",
+											  task->jobid, task->taskid)));
+
+							task->status = BACKGROUND_TASK_STATUS_CANCELLED;
+							monitorExecutionState = TaskEnded;
+						}
+						else if (hadError)
+						{
+							ereport(LOG, (errmsg("task jobid/taskid failed: %ld/%ld",
+												 task->jobid, task->taskid)));
+
+							monitorExecutionState = TaskHadError;
+						}
+						else if (mq_res == SHM_MQ_WOULD_BLOCK)
+						{
+							/*
+							 * still running the task but had no output for current iteration
```

Comment:
it's not necessarily true that it had no output right? It might have logged/noticed/warned stuff.

Reply:
I removed the 'had no output' part but the shared memory queue has nothing to read if it returned SHM_MQ_WOULD_BLOCK.